Added Attachments support for LRSs

FishHooks opened the pull request:
Rebased with current master and fixed tests to support version 1.0.2

Addresses #70 

This PR includes the following:

- Attachments can now be added to statements and sent to the LRS

- The attachments' content is sent with the statement but not serialized with the rest of the statement's JSON

- The LRS sends the put request in the correct format if the statement has an attachment (and the correct `Content-Type`)

- If the attachment flag is set to `true` in `queryStatements()`, the response is formatted correctly (previously `queryStatements()` did not handle multipart requests, and would therefore not return the attachments' content)

brianjmiller commented:
It would be good to test the additional methods of LRS, and to have a test that sends an attachment, one that retrieves a statement via id, and one that query's statements.

brianjmiller commented on src/Attachment.js:
```diff
@@ -0,0 +1,176 @@
+/*
+    Copyright 2014 Rustici Software
```
New files should have updated year, 2016.

brianjmiller commented on src/Attachment.js:
```diff
@@ -0,0 +1,176 @@
+            }
+            if (cfg.hasOwnProperty("fileUrl") && typeof cfg.fileUrl !== "undefined" && cfg.fileUrl !== null) {
+                this.fileUrl = cfg.fileUrl;
+            }
```
Can't these two be `directProps`?

brianjmiller commented on src/Attachment.js:
```diff
@@ -0,0 +1,176 @@
+                }
+            }
+
+            if (cfg.hasOwnProperty("content") && typeof cfg.content !== "undefined" && cfg.content !== null) {
```
Not sure why we are checking both `hasOwnProperty` and the `typeof`, generally the latter is used to check to see if a property exists. Shouldn't need both.

brianjmiller commented on src/Attachment.js:
```diff
@@ -0,0 +1,176 @@
+        @type Object
+        */
+        this.description = null;
+
```
Should include `content` in the prototype since it will sometimes be set, then can just check for `null` to see if it *has* been set.

brianjmiller commented on src/Attachment.js:
```diff
@@ -0,0 +1,176 @@
+                    sha2: this.sha2,
+                    usageType: this.usageType
+                };
+                if (this.hasOwnProperty("content") && typeof this.content !== "undefined" && this.content !== null) {
```
`this` will *always* have `content` based on the object's prototype based on comment above, this can be reduced to just check for null.

brianjmiller commented on src/Attachment.js:
```diff
@@ -0,0 +1,176 @@
+                    length: this.length,
+                    sha2: this.sha2,
+                    usageType: this.usageType
+                };
```
`description` and `fileUrl` are optional, need a `null` check to not include them.

brianjmiller commented on src/LRS.js:
```diff
@@ -158,6 +158,39 @@ TinCan client library
+        requests where the statement has an attachment
+
+        @method getBoundary
+        @private
```
JS doesn't have actual private/public, etc. so properties/methods that are "private"ish should start with `_`, please rename these.

brianjmiller commented on src/LRS.js:
```diff
@@ -158,6 +158,39 @@ TinCan client library
+        @private
+        */
+        createStatementSegment: function (boundary, statement) {
+            return "--" + boundary + "\r\n" + "Content-Type:application/json" + "\r\n\r\n" + JSON.stringify(statement) + "\r\n";
```
Per RFC a space is preferred after header names and before the value.

brianjmiller commented on src/LRS.js:
```diff
@@ -310,7 +343,11 @@ TinCan client library
+                boundary,
+                flag = true,
+                attachmentContent = [],
+                a;
```
Getting *really* nitpicky now :-), but `i` is more canonical in the lib (AFAICT all for loops use it), so let's switch for consistency.

brianjmiller commented on src/LRS.js:
```diff
@@ -310,7 +343,11 @@ TinCan client library
-                versionedStatement;
+                versionedStatement,
+                boundary,
+                flag = true,
```
`flag` for what? Better name please.

brianjmiller commented on src/LRS.js:
```diff
@@ -341,13 +378,43 @@ TinCan client library
+                    }
+                };
+                for (a = 0; a < versionedStatement.attachments.length; a += 1) {
+                    if (typeof versionedStatement.attachments[a].content !== "undefined" && versionedStatement.attachments[a].content !== null) {
```
Should be just `null` check after above.

brianjmiller commented on src/LRS.js:
```diff
@@ -341,13 +378,43 @@ TinCan client library
+                for (a = 0; a < versionedStatement.attachments.length; a += 1) {
+                    if (typeof versionedStatement.attachments[a].content !== "undefined" && versionedStatement.attachments[a].content !== null) {
+                        attachmentContent.push(versionedStatement.attachments[a].content);
+                        delete versionedStatement.attachments[a].content;
```
This begs the question of whether the versioned statement should *ever* have the `content` set, since it can't be used in that way. It probably shouldn't, and instead we use the index of the versioned attachments to index into the original statement to check for content. See PHP for an example.

brianjmiller commented on src/LRS.js:
```diff
@@ -341,13 +378,43 @@ TinCan client library
+                    requestCfg.data += "--" + boundary + "--";
+                }
+            }
+            else if (flag) {
```
When would `flag` be `false`?

brianjmiller commented on src/LRS.js:
```diff
@@ -542,6 +609,46 @@ TinCan client library
+                    i += 1;
+                    content = lines[i];
+                    for (x = 0; x < stmts.statements.length; x += 1) {
+                        if (stmts.statements[x].hasOwnProperty("attachments") && stmts.statements[x].attachments !== null) {
```
Should be `null` only check.

brianjmiller commented on src/LRS.js:
```diff
@@ -542,6 +609,46 @@ TinCan client library
+                }
+                else if (lines[i] === "") {
+                    i += 1;
+                    content = lines[i];
```
What happens if content is on more than one line?

brianjmiller commented on src/LRS.js:
```diff
@@ -542,6 +609,46 @@ TinCan client library
+
+            for (i = 0; i < lines.length; i += 1) {
+                if (lines[i].startsWith("X-Experience-API-Hash")) {
+                    hashToMatch = lines[i].split(":")[1];
```
Not sure this is sufficient to parse headers. Can we avoid parsing the headers ourselves? At the very least this probably breaks on whitespace.

brianjmiller commented on src/LRS.js:
```diff
@@ -604,10 +711,29 @@ TinCan client library
+                        }
+                        else {
+                            for (i = 0; i < responseLines.length; i += 1) {
+                                if (responseLines[i].startsWith("Content-Type:application/json")) {
```
In sufficient for header detection.

brianjmiller commented on src/LRS.js:
```diff
@@ -604,10 +711,29 @@ TinCan client library
+                        else {
+                            for (i = 0; i < responseLines.length; i += 1) {
+                                if (responseLines[i].startsWith("Content-Type:application/json")) {
+                                    i += 2;
```
?? No guarantee content type is the last header.

brianjmiller commented on src/Statement.js:
```diff
@@ -179,7 +186,8 @@ TinCan client library
                     "voided"
-                ];
+                ],
+                a;
```
prefer `i` for single indexing.

brianjmiller commented on src/Statement.js:
```diff
@@ -278,6 +286,16 @@ TinCan client library
                     this.context = new TinCan.Context (cfg.context);
                 }
             }
+            if (cfg.hasOwnProperty("attachments") && cfg.attachments !== null && typeof cfg.attachments !== "undefined") {
```
Don't need both checks.

brianjmiller commented on src/Statement.js:
```diff
@@ -278,6 +286,16 @@ TinCan client library
+                for (a = 0; a < cfg.attachments.length; a += 1) {
+                    if (!(cfg.attachments[a] instanceof TinCan.Attachment)) {
+                        this.attachments.push(new TinCan.Attachment (cfg.attachments[a]));
+                    } else {
```
Eek, cuddled else!

brianjmiller commented on src/LRS.js:
```diff
@@ -618,7 +744,7 @@ TinCan client library
             requestResult.config = requestCfg;
 
-            if (! callbackWrapper) {
+            if (!callbackWrapper) {
```
Should be a space here.

brianjmiller commented on src/LRS.js:
```diff
@@ -604,10 +711,29 @@ TinCan client library
 
                     if (err === null) {
-                        result = TinCan.StatementsResult.fromJSON(xhr.responseText);
+                        if (!cfg.params.attachments) {
```
Space after operator.

brianjmiller commented on src/LRS.js:
```diff
@@ -341,13 +378,43 @@ TinCan client library
+                    flag = false;
+                    requestCfg.headers["Content-Type"] = "multipart/mixed; boundary=" + boundary;
+                }
+                if (!flag) {
```
Space after operator.

brianjmiller commented on src/Statement.js:
```diff
@@ -278,6 +286,16 @@ TinCan client library
+            if (cfg.hasOwnProperty("attachments") && cfg.attachments !== null && typeof cfg.attachments !== "undefined") {
+                this.attachments = [];
+                for (a = 0; a < cfg.attachments.length; a += 1) {
+                    if (!(cfg.attachments[a] instanceof TinCan.Attachment)) {
```
Space after operator.

brianjmiller commented on src/Statement.js:
```diff
@@ -321,7 +339,9 @@ TinCan client library
-                i;
+                a,
+                i,
+                temp;
```
`temp` should be unnecessary, use method chaining.

brianjmiller replied:
Alternatively, always set it (as a better name) and drop the `.push` redundancy.

brianjmiller commented on test/js/unit/Attachment.js:
```diff
@@ -0,0 +1,167 @@
+/*
+    Copyright 2014 Rustici Software
```
+1

brianjmiller commented on test/js/unit/Attachment.js:
```diff
@@ -0,0 +1,167 @@
+            }
+
+            strictEqual(obj.LOG_SRC, "Attachment", "object property LOG_SRC initial value");
+            //strictEqual(obj.objectType, "Attachment", "object property objectType initial value");
```
Remove.

brianjmiller commented on test/js/unit/Attachment.js:
```diff
@@ -0,0 +1,167 @@
+                    }
+                }
+            ],
+                i,
```
Something is odd with this indenting.

brianjmiller commented on test/js/unit/Statement.js:
```diff
@@ -74,6 +74,9 @@
                 context = new TinCan.Context(),
                 authority = new TinCan.Agent({ mbox: "tincanjs-test-authority@tincanapi.com" }),
+                attachments = [
+                        new TinCan.Attachment({ usageType: "http://id.tincanapi.com/attachment/test-attachment" })
```
Whitespace.

brianjmiller commented on test/single/Attachment.html:
```diff
@@ -0,0 +1,36 @@
+<!DOCTYPE html>
+<!--
+    Copyright 2014 Rustici Software
```
+1

FishHooks commented:
This last commit adds tests for the new LRS methods, as well as sending and receiving statements with attachments.

Additionally, I've removed all manual header parsing and used the built-in methods of XHR to handle the response headers.

brianjmiller commented on src/LRS.js:
```diff
@@ -456,12 +449,34 @@ TinCan client library
             if (typeof cfg.callback !== "undefined") {
                 callbackWrapper = function (err, xhr) {
-                    var result = xhr;
+                    var result = xhr.responseText,
```
Changing the initialization of `result` here breaks the public interface contract in the case of errors. (Which I honestly thought would have been caught by the unit tests :-(). See `cfg.callback(err, result)` below.

brianjmiller commented on src/LRS.js:
```diff
@@ -456,12 +449,34 @@ TinCan client library
+
+                            for (i = 2; i < result.length; i += 1) {
+                                section = result[i].split("\r\n\r\n");
+                                statement = TinCan.LRS.prototype._assignAttachmentContent({statements: [statement]}, section[0].split("\r\n"), section[1].trim()).statements[0];
```
Shouldn't call this method on the prototype without doing so through the instance itself. It should be made either local and not available externally, "static" on the "class", or make the instance available likely via a closure and call it as an instance method.

brianjmiller commented on src/LRS.js:
```diff
@@ -711,25 +712,25 @@ TinCan client library
                 callbackWrapper = function (err, xhr) {
-                    var result = xhr,
-                        responseLines = xhr.responseText.split("\r\n"),
+                    var result = xhr.responseText,
```
+1 See comment above.

brianjmiller commented on src/LRS.js:
```diff
@@ -711,25 +712,25 @@ TinCan client library
+
+                            for (i = 2; i < result.length; i += 1) {
+                                section = result[i].split("\r\n\r\n");
+                                statements = TinCan.LRS.prototype._assignAttachmentContent(statements, section[0].split("\r\n"), section[1].trim());
```
+1 See comment above.

brianjmiller commented on src/LRS.js:
```diff
@@ -887,6 +888,42 @@ TinCan client library
+
+        @method _assignAttachmentContent
+        @private
+        @param {Array} [stmts] Array of TinCan Statement JSON objects
```
"Tin Can" or "TinCan.Statement" but not "TinCan".

brianjmiller commented on src/LRS.js:
```diff
@@ -887,6 +888,42 @@ TinCan client library
+        @param {Array} [stmts] Array of TinCan Statement JSON objects
+        @param {Array} [headers] Array containing section headers from the HTTP response
+        @param {String} [content] The content to place into its attachment
+        @return {Array} Array of TinCan Statement JSON objects with correctly assigned attachment content
```
+1

brianjmiller commented on src/LRS.js:
```diff
@@ -887,6 +888,42 @@ TinCan client library
+
+            for (i = 0; i < headers.length; i += 1) {
+                if (headers[i].indexOf("X-Experience-API-Hash") >= 0) {
+                    hashToMatch = headers[i].split(":")[1].trim();
```
Still header parsing. What happens if there is whitespace around the ":"?

brianjmiller commented on src/LRS.js:
```diff
@@ -887,6 +888,42 @@ TinCan client library
+            }
+
+            for (i = 0; i < stmts.statements.length; i += 1) {
+                if (stmts.statements[i].hasOwnProperty("attachments") && stmts.statements[i].attachments !== null) {
```
No need to check if a statement has an attachments property, it always should.

brianjmiller replied:
Unless these are the JSON objects, have they already been turned into `TinCan.Statement` objects?

FishHooks replied:
@brianjmiller At this point they're just JSON, and then I turn them into to `TinCan.Statement` objects after this function returns.

brianjmiller commented on src/LRS.js:
```diff
@@ -887,6 +888,42 @@ TinCan client library
+        @param {String} [content] The content to place into its attachment
+        @return {Array} Array of TinCan Statement JSON objects with correctly assigned attachment content
+        */
+        _assignAttachmentContent: function (stmts, headers, content) {
```
I wonder if it makes more sense to collect all attachments based on hash, create a map of them, then associate with the statements via the map which should be faster because you'll only run through the statements + attachment lists once, with the current implementation you have to loop over the full list of attachments for each attachment. Granted the lists are small, but no reason to include the overhead AFAICT.

brianjmiller commented on src/Statement.js:
```diff
@@ -398,6 +395,29 @@ TinCan client library
+            }
+            else {
+                for (i = 0; i < this.attachments.length; i += 1) {
+                    if (this.attachments[i].hasOwnProperty("content") && this.attachments[i].content !== null) {
```
An `TinCan.Attachment` always has "content" as a property.

brianjmiller commented on test/js/unit/LRS.js:
```diff
@@ -189,6 +189,45 @@
+        "_createStatementSegment",
+        function () {
+            var testString = "--testBoundary\r\nContent-Type: application/json\r\n\r\n\"test\"\r\n";
+            deepEqual(TinCan.LRS.prototype._createStatementSegment("testBoundary", "test"), testString, "Statement segment created correctly");
```
+1 prototype usage.

brianjmiller commented on test/js/unit/LRS.js:
```diff
@@ -189,6 +189,45 @@
+        "_createAttachmentSegment",
+        function () {
+            var testString = "--testBoundary\r\nContent-Type: text/plain\r\nContent-Transfer-Encoding: binary\r\nX-Experience-API-Hash: testHash\r\n\r\ntest\r\n";
+            deepEqual(TinCan.LRS.prototype._createAttachmentSegment("testBoundary", "test", "testHash", "text/plain"), testString, "Statement segment created correctly");
```
+1 prototype usage.

brianjmiller commented on test/js/unit/TinCan-async.js:
```diff
@@ -180,7 +242,7 @@
                         },
                         target: {
-                            id: "http://tincanapi.com/TinCanJS/Test/TinCan_getStatement/async/" + v
+                            id: "http://tincanapi.com/TinCanJS/Test/TinCan_getStatements/async/" + v
```
This is testing `getStatement`, shouldn't change?

brianjmiller commented on test/js/unit/TinCan-async.js:
```diff
@@ -229,11 +291,139 @@
+                            id: "http://adlnet.gov/expapi/verbs/attempted"
+                        },
+                        target: {
+                            id: "http://tincanapi.com/TinCanJS/Test/TinCan_getStatements/async/" + v
```
+1 testing `getStatement`?

brianjmiller commented on test/js/unit/TinCan-async.js:
```diff
@@ -229,11 +291,139 @@
+                            id: "http://adlnet.gov/expapi/verbs/attempted"
+                        },
+                        target: {
+                            id: "http://tincanapi.com/TinCanJS/Test/TinCan_getStatement/async/" + v
```
This time testing `getStatements`?

brianjmiller commented on test/js/unit/TinCan-async.js:
```diff
@@ -229,11 +291,139 @@
+                                statement.authority = null;
+
+                                deepEqual(err, null, "callback: err argument (" + v + ")");
+                                deepEqual(sentStatement.attachments[0], new TinCan.Attachment(statement.attachments[0]), "callback: statement matches (" + v + ")");
```
Shouldn't  the statement coming back already have a `TinCan.Attachment`, why construct it here?

brianjmiller commented on test/js/unit/TinCan-async.js:
```diff
@@ -229,11 +291,139 @@
+                                callback: function (err, statement) {
+                                    start();
+                                    deepEqual(err, null, "callback: err argument (" + v + ")");
+                                    deepEqual(sentStatement.attachments[0], new TinCan.Attachment(statement.statements[0].attachments[0]), "callback: statement matches (" + v + ")");
```
+1 to construction question.

brianjmiller commented on src/LRS.js:
```diff
@@ -454,27 +455,35 @@ TinCan client library
             if (typeof cfg.callback !== "undefined") {
                 callbackWrapper = function (err, xhr) {
-                    var result = xhr.responseText,
+                    var result,
```
This didn't put the original back so is still broken, `result` needs to be initialized to `xhr` so that it carries through in the callback in the case of errors.

brianjmiller commented on src/LRS.js:
```diff
@@ -454,27 +455,35 @@ TinCan client library
                             for (i = 2; i < result.length; i += 1) {
                                 section = result[i].split("\r\n\r\n");
-                                statement = TinCan.LRS.prototype._assignAttachmentContent({statements: [statement]}, section[0].split("\r\n"), section[1].trim()).statements[0];
+                                headers = section[0].split("\r\n");
```
I'm not convinced any of these `split`s will actually work as intended. They'll handle the simple case correctly, but will fall down on either complex headers, and/or bodies that include non-normal whitespace. Minimally they need to include the second limit argument, specifying that the split should only occur once. For instance, headers can be split across lines or bodies might include multiple line breaks in a row, hence the separation of header from body + boundary.

brianjmiller commented on src/LRS.js:
```diff
@@ -454,27 +455,35 @@ TinCan client library
+                                headers = section[0].split("\r\n");
+                                for (i = 0; i < headers.length; i += 1) {
+                                    if (headers[i].indexOf("X-Experience-API-Hash") >= 0) {
+                                        attachmentMap[TinCan.Utils.getXAPIHashFromHeader(headers[i])] = section[1].replace(/^\s+|\s+$/g,"");
```
This `.replace` is the reason I was concerned about the above as well, this will effectively trim valid whitespace from the body of the content when it shouldn't. We can't rely on the body of the message part not having leading/trailing whitespace, which is the point of the boundary conditions. We likely need to set up some more advanced test cases to make sure we not mangling the attachments.

brianjmiller commented on src/LRS.js:
```diff
@@ -893,34 +917,25 @@ TinCan client library
             }
 
-            return stmts;
+            return (stmts.length === 1 ? stmts[0] : stmts);
```
I think this should be a void return, no reason to return the stmts since they are modified by the function itself.

brianjmiller commented on src/LRS.js:
```diff
@@ -893,34 +917,25 @@ TinCan client library
-        @param {String} [content] The content to place into its attachment
-        @return {Array} Array of TinCan Statement JSON objects with correctly assigned attachment content
+        @param {Array} [stmts] Array of TinCan.Statement JSON objects
+        @param {Hashmap} [content] Hashmap of the content to place into its attachment
```
Is "Hashmap" a valid parameter type in these docs? I think we can stick with object in the JS world.

brianjmiller commented on src/LRS.js:
```diff
@@ -893,34 +917,25 @@ TinCan client library
-        _assignAttachmentContent: function (stmts, headers, content) {
-            var hashToMatch,
-                i,
+        _assignAttachmentContent: function (stmts, content) {
```
"content" should probably be renamed, "attachments" or even "attachmentMap" is more clear.

brianjmiller commented:
Based on the other PR going in, this one needs to be merged/rebased with master again to resolve the conflicts. I'd probably lean rebase.

FishHooks commented:
Alright. It's refactored, rebased, and retested. I ported the multipart response parsing helper methods straight over from PHP, so it should have everything. The only thing I removed was that PHP had it set up to put the status in the first element of the multipart sections array, so I removed that (I'm guessing it was there because of the way PHP handles the response).

tseabrooks commented:
@brianjmiller We need to have this merged.

Thanks, TJ

brianjmiller commented:
Closing in favor of #133.
